src/libstd/sys/unix/fs.rs

```diff
@@ -732,17 +791,126 @@ pub fn canonicalize(p: &Path) -> io::Result<PathBuf> {
 }
 
 pub fn copy(from: &Path, to: &Path) -> io::Result<u64> {
-    use fs::{File, set_permissions};
-    if !from.is_file() {
+    use fs::OpenOptions;
+    let mut from_opts = OpenOptions::new();
+    let mut reader = try!(from_opts.read(true).open(&from));
+    let metadata = try!(reader.metadata());
+    let attr = metadata.as_inner();
+    if attr.file_type().is_file() == false {
         return Err(Error::new(ErrorKind::InvalidInput,
                               "the source path is not an existing regular file"))
     }
 
-    let mut reader = try!(File::open(from));
-    let mut writer = try!(File::create(to));
-    let perm = try!(reader.metadata()).permissions();
+    let mut to_opts = OpenOptions::new();
+    let mut writer = try!(to_opts.write(true)
+                                 .create(true)
+                                 .truncate(true)
+                                 .mode(attr.perm().mode())
```

Comment:
Can you create a writable version of a read-only file? (I'm just not sure if that works here)